### PR TITLE
date updates

### PR DESCRIPTION
**Pull request by hardy925**
## Fixes

_Please link to the issue this PR addresses_

## Description

better ISO 8106 support

## Things to look at

- [x] Test coverage
- [x] Code Style
- [x] Documentation (READMEs etc)

### CODE REVIEW

File: from_json/date.ts [hunk 1]
```diff
@@ -11,13 +11,78 @@ export function createDateStrategy(regex: RegExp): FromJsonStrategy {
...
+}
+
+/** Test a string for a full ISO 8601 Date */
+export function iso8601Date(input: JsonValue): any {
+  var iso =
+    /^(\d{4})(?:-?W(\d+)(?:-?(\d+)D?)?|(?:-(\d+))?-(\d+))(?:[T ](\d+):(\d+)(?::(\d+)(?:\.(\d+))?)?)?(?:Z(-?\d*))?$/;
+  // throw error on bad data?
+  if (typeof input !== "string" || !iso.exec(input)) {
+    return input;
+  }
```
Comment (ChrisDufourMB) [inline from_json/date.ts]: I think we should throw an error here instead of returning the input, or at least return `undefined`.

If we eventually add pre-parse validation we should have a proper way of validating this and handling errors, but for now, I think we should just try to avoid returning different types when errors occur.

Another option would be to simply return an `Invalid Date` type date, not sure how we would initialize that though.

File: from_json/date.ts [hunk 2]
```diff
@@ -11,13 +11,78 @@ export function createDateStrategy(regex: RegExp): FromJsonStrategy {
...
+  year: number;
+  week: number;
+  weekday: number;
+  timezone: number;
+}
+
+/** Test a string for a full ISO 8601 Date */
+export function iso8601Date(input: JsonValue): any {
+  var iso =
+    /^(\d{4})(?:-?W(\d+)(?:-?(\d+)D?)?|(?:-(\d+))?-(\d+))(?:[T ](\d+):(\d+)(?::(\d+)(?:\.(\d+))?)?)?(?:Z(-?\d*))?$/;
```
Comment (ChrisDufourMB) [inline from_json/date.ts]: A comment explaining what this regex does would be appreciated

File: from_json/date.ts [hunk 3]
```diff
@@ -11,13 +11,78 @@ export function createDateStrategy(regex: RegExp): FromJsonStrategy {
...
+  // throw error on bad data?
+  if (typeof input !== "string" || !iso.exec(input)) {
+    return input;
+  }
+  // first match is the whole pattern which we do not need `,` skips it
+  // we're also skipping week and weekday
+  let [
```
Comment (ChrisDufourMB) [inline from_json/date.ts]: Could you add a quick comment with an example array? There are a lot of unused variables, so it's a bit difficult to understand what the format coming out of `input.match(iso)` would be like

File: from_json/date.ts [hunk 4]
```diff
@@ -11,13 +11,78 @@ export function createDateStrategy(regex: RegExp): FromJsonStrategy {
...
+      year,
+      ,
+      ,
+      month,
+      day,
+      hour,
+      minute,
+      second,
+      millisecond,
+      zone,
```
Comment (ChrisDufourMB) [inline from_json/date.ts]: How does the zone support work? IIRC zones can be something like this:`+02:30`, which can't be cleanly converted to a number

Also note that a zone like this: `Z+0930` is not + 930 hours, but +9 hours, 30 minutes (basically it's just 09:30 without the `:`)

File: from_json/date.ts [hunk 5]
```diff
@@ -11,13 +11,78 @@ export function createDateStrategy(regex: RegExp): FromJsonStrategy {
...
+    ] = input.match(iso)?.map(Number) || [],
+    timezone = parseTimezone(0);
+
+  if (Number.isNaN(year)) {
+    return input;
```
Comment (ChrisDufourMB) [inline from_json/date.ts]: See other comment on error handling, same applies here

File: from_json/date.ts [hunk 6]
```diff
@@ -11,13 +11,78 @@ export function createDateStrategy(regex: RegExp): FromJsonStrategy {
...
- *  YYYY-MM-DDThh:mm:ss.sTZD
- * @see: https://www.w3.org/TR/NOTE-datetime
- * @type {RegExp}
- */
-export const ISODateFromJson = createDateStrategy(
-  /^\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d(\.\d+)?(([+-]\d\d:\d\d)|Z)?$/i,
-);
+/** converts timezone to minutes */
+function parseTimezone(zone: number): number {
+  return zone * 60 - Number.parseInt(`-${new Date().getTimezoneOffset()}`);
```
Comment (ChrisDufourMB) [inline from_json/date.ts]: `new Date().getTimezoneOffset()` returns a number, it seems a bit strange to me that we would be converting it to a string, then parsing it back to a number like this. Parsing it as a number, then subtracting it seems a bit redundant, would it make more sense to simply add it instead?

Does this work in negative time zones?

File: from_json/date_test.ts [hunk 7]
```diff
@@ -1,72 +1,81 @@
...
-    assertEquals(testObj.date.getFullYear(), 2020);
-    assert(!(testObj.not_a_date instanceof Date));
-    assertEquals(testObj.not_a_date, "Hello world");
+    assertEquals(testObj.date.toISOString(), "2020-12-31T12:00:00.000Z");
   },
 });
 
 test({
-  name: "ISODateFromJson parses ISO dates - 2020-12-31T23:00:00",
+  name: "iso8601Date parses with added timezone - 2020-12-31T12:00:00Z-6",
```
Comment (ChrisDufourMB) [inline from_json/date_test.ts]: I don't think single-digit offsets are technically part of the ISO standard, I believe that a proper year offset needs to be left-padded with 0s

**Review by ChrisDufourMB — COMMENTED** [review]
This doesn't appear to support all the different kinds of timezones that could be part of an ISO date, what degree of support are you looking to implement? I feel like we will need more tests cases for all the different variants and edge cases that can happen

Here is the test I tried, for reference
```ts

test({
  name: "iso8601Date parses with added timezone - 2020-12-31T12:00:00Z-06:00",
  fn() {
    const testJson = `{"date":"2020-12-31T06:00:00Z-06:00"}`;
    const testObj = JSON.parse(testJson, (_, v) => iso8601Date(v));
    assert(testObj.date instanceof Date);
    assertEquals(testObj.date.toISOString(), "2020-12-31T12:00:00.000Z-06:00");
  },
});
```

File: from_json/date.ts (hunk 1 above)
Comment (ms1111) [reply from_json/date.ts]: Throwing an Error would be kind of how Jackson handles invalid input in Java.

File: from_json/date_test.ts [hunk 8]
```diff
@@ -1,72 +1,81 @@
...
+});
+
+test({
+  name: "iso8601Date parses `2020-366` - leap year",
+  fn() {
+    const testJson = `{"date":"2020-366"}`;
+    const testObj = JSON.parse(testJson, (_, v) => iso8601Date(v));
+    assert(testObj.date instanceof Date);
+    assertEquals(testObj.date.toISOString(), "2020-12-31T00:00:00.000Z");
   },
```
Comment (ms1111) [inline from_json/date_test.ts:84]: Can I suggest a test to check that YYYY-MM-DD is deserialized as midnight local time? (I think this test will pass, looking at the code)

This is how we'll want it for our date types. (Code that uses isoDateStringToLocalDate() now)

Here is how we tested it in date-utils.spec.ts:
```
    // 1990-12-11
    const localDateMidnight = new Date(1990, 11, 11);
...
expect(isoDateStringToLocalDate("1990-12-11").getTime()).toEqual(localDateMidnight.getTime());
```

**Review by ms1111 — COMMENTED** [review]
Would you also want to have a serializer in the library to render a local date as an ISO date string (YYYY-MM-DD) in toJson()? Or would you want apps to provide that themselves (like our localDateToIsoDateString() function)

Here's the test case for localDateToIsoDateString() btw:
```
    // 1990-12-11
    const localDateMidnight = new Date(1990, 11, 11);

expect(localDateToIsoDateString(localDateMidnight)).toEqual("1990-12-11");
```

code:
```
export function localDateToIsoDateString(date: Date): string {
    const year = date.getFullYear();
    const month = date.getMonth() + 1;
    const day = date.getDate();
    return `${formatISOYear(year)}-${padLeft(month, 2)}-${padLeft(day, 2)}`;
}
```

File: from_json/date.ts (hunk 6 above)
Comment (hardy925) [reply from_json/date.ts]: Removing this in place for `Date.UTC`

File: from_json/date.ts (hunk 1 above)
Comment (hardy925) [reply from_json/date.ts]: done

File: from_json/date.ts (hunk 4 above)
Comment (hardy925) [reply from_json/date.ts]: remove most of this, and I used `match` ... exec would be better for this

File: from_json/date.ts (hunk 5 above)
Comment (hardy925) [reply from_json/date.ts]: done

File: from_json/date_test.ts (hunk 8 above)
Comment (hardy925) [reply from_json/date_test.ts:84]: This depends on the string, "1990-12-11" will have zones applied to it,  where "1990/12/11" will not. I am going to divert this so it is up to authors to make the strategies they want now.

**Comment by hardy925** [comment]
@ms1111 I think we'll go the route of having authors write their strategies, the two we provide for dates is more than enough IMHO.
